### PR TITLE
Add support for item() and nonzero() codegen in Inductor

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #110195
* __->__ #109893

This is another version of
https://github.com/pytorch/pytorch/pull/109262 that I think is more
harmonious with inductor design.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @ngimel

### CODE REVIEW

**Comment by ezyang** [comment]
Discussion from @Chillee is that he is roughly ok with the approach, but he wants us to setup dependencies between DynamicScalar and the downstream usages of it (i.e., whenever the unbacked symint shows up.) This should be done in the scheduler.

File: torch/_inductor/codegen/cpp.py [hunk 1]
```diff
@@ -1415,7 +1420,12 @@ def decide_parallel_depth(self, ranges, threads):
         par = 1
         depth = 0
         for expr in ranges:
-            hint = V.graph.sizevars.size_hint(expr)
+            sym_hint = V.graph.sizevars.symbolic_hint(expr)
+            if not isinstance(sym_hint, (int, sympy.Integer)):
+                # Make a guess
+                hint = 8192
+            else:
+                hint = int(sym_hint)
```
Comment (jansel) [inline torch/_inductor/codegen/cpp.py:1428]: With the changes to size_hint above, is this change still needed?

File: torch/_inductor/ir.py [hunk 2]
```diff
@@ -2417,7 +2436,7 @@ def decide_layout(self):
         pass
 
     def get_alias_names(self):
-        if isinstance(self.layout, AliasedLayout):
+        if isinstance(self.layout, (AliasedLayout)):
```
Comment (jansel) [inline torch/_inductor/ir.py:2439]: revert?

**Review by jansel — APPROVED** [review]

**Comment by ezyang** [comment]
reupping for review since the dep addition changes are fairly major

File: torch/_inductor/codegen/cpp.py (hunk 1 above)
Comment (ezyang) [reply torch/_inductor/codegen/cpp.py:1432]: yeah, because we need an actual integer to run the rest of this code, IIUC

Comment (jansel) [reply torch/_inductor/codegen/cpp.py:1432]: Couldn't we just call size_hint() because it looks like the same code?

File: torch/_inductor/ir.py [hunk 3]
```diff
@@ -2511,6 +2555,26 @@ def get_read_writes(self):
...
+        #
+        # For now, I'm just yoloing with the size of the buffer.  Not sure if
+        # it is enough.
+        #
+        # One thing you might wonder is if this is enough for a ComputedBuffer
+        # denoting a reduction over i0.  Empirically, it is enough, but for an
+        # unusual reason: we only need accurate dependencies for item() call,
+        # but it's impossible to end up with a reduction over i0 from an
+        # item() call without a regular non-reduction buffer first.
+        return free_unbacked_symbols(self.get_size())
```
Comment (jansel) [inline torch/_inductor/ir.py:2576]: stride/offset needs checking?

File: torch/_inductor/ir.py [hunk 4]
```diff
@@ -3305,6 +3369,31 @@ def canonicalize(self):
...
+        def maybe_free_unbacked_symbols(s):
+            if isinstance(s, (SymTypes, sympy.Expr)):
+                return free_unbacked_symbols(s)
+            elif isinstance(s, (tuple, list)):
+                r = set()
+                for t in s:
+                    r |= maybe_free_unbacked_symbols(t)
+                return r
+            else:
+                return set()
```
Comment (jansel) [inline torch/_inductor/ir.py:3388]: Some sort of assert based on type to ensure we aren't missing something.

**Review by jansel — DISMISSED** [review]

File: torch/_inductor/codegen/cpp.py (hunk 1 above)
Comment (ezyang) [reply torch/_inductor/codegen/cpp.py:1432]: I misunderstood your comment originally.

The `size_hint` above is defined in `CppKernel` and hard coded to `sympy_product(self.call-ranges)`. So no, I cannot "just call size_hint()" to hit the same code, at minimum there needs to be a refactor.

But we should not do this refactor. There is no *intrinsic* reason why the hints here should actually be the same. They happen to be the same because I am just randomly picking numbers. In fact, they probably should not be the same: the hint in `size_hint` is the product of all call ranges, which should be guessed to be larger than the hint for a single range. I am hoping we can revisit the hint selection later when we can do things like (1) have upper bounds on the size of the unbacked symint, which can be used to drive the hint or (2) maybe come up with some better heuristics armed with actual slowness that we're looking to speed up.

File: torch/_inductor/ir.py (hunk 4 above)
Comment (ezyang) [reply torch/_inductor/ir.py:3388]: I'm going to appeal to pre-existing condition here. As the comment notes, this is modeled off of val_to_arg_str which has this structure:

```
    def val_to_arg_str(self, s):
        if isinstance(s, SymTypes):
            return pexpr(sympy.expand(repr(s)))
        elif isinstance(s, sympy.Expr):
            return pexpr(s)
        elif isinstance(s, (tuple, list)):

            @dataclasses.dataclass
            class Shim:
                ref: Any

                def __repr__(self):
                    return self.ref

            return repr(type(s)(Shim(self.val_to_arg_str(a)) for a in s))
        elif isinstance(s, torch._ops.OpOverload):
            return _get_qualified_name(s)
        else:
            return repr(s)
```

This function has a universal fallthrough; so shall we.

I agree it would be better if we could explicate what the set of valid arguments is. But I am not too sure what this set is. Naively, I would have guessed it to be the set of valid types that arguments in JIT schema can take. But there's a special case for OpOverload in here! So... it also includes valid higher order op arguments? But higher order op arguments can literally be anything...

File: torch/_inductor/ir.py (hunk 3 above)
Comment (ezyang) [reply torch/_inductor/ir.py:2576]: bah, fine, I'll implement this (I have a hard time imagining anyone who will actually have dynamic stride/offset in this way lol)

**Comment by ezyang** [comment]
stride support pretty involved, check https://github.com/pytorch/pytorch/pull/109893/commits/db98cca27f78faa27160f8e5d01434d457848032

File: test/inductor/test_torchinductor_dynamic_shapes.py [hunk 5]
```diff
@@ -155,6 +157,119 @@ def fn(x, a):
...
+
+            @torch.compile(fullgraph=True)
+            def f(x, r):
+                y = x.item()
+                return torch.ops.test.foo(r, y)
+
+            f(torch.tensor([3], device=device), torch.randn(10, device=device))
+
+        finally:
+            custom_ops._destroy("test::foo")
```
Comment (ezyang) [inline test/inductor/test_torchinductor_dynamic_shapes.py:236]: @zou3519 take a look at these test cases (here and below) and tell me how I'm doing it wrong :P

Comment (ezyang) [reply test/inductor/test_torchinductor_dynamic_shapes.py:236]: yeah, essentially the regs here are annoyingly asymmetric but as best I could tell there was no way to actually do it symmetrically

File: torch/_inductor/scheduler.py [hunk 6]
```diff
@@ -1225,7 +1225,26 @@ def add_user(used_by_name, user_node, can_inplace=False, is_weak=False):
...
+                unbacked_symbol_to_origin_node[s] = node
+
+            # if a kernel takes unbacked symints, register dependencies
+            for s in node.node.get_unbacked_symbol_uses():
+                # NB: This is not actually a mutation dep, but we do need to
+                # record this ordering dependency
+                assert (
+                    s in unbacked_symbol_to_origin_node
+                ), f"{s} not in {unbacked_symbol_to_origin_node}"
+                node.add_mutation_dep(
```
Comment (ezyang) [inline torch/_inductor/scheduler.py:1244]: @jansel says maybe we should rename this

**Review by jansel — APPROVED** [review]

**Review by eellison — COMMENTED** [review]
Can you add a call to `V.graph.disable_cudagraphs = True` somewhere here ?

**Comment by ezyang** [comment]
@eellison i just added _local_scalar_dense to the forbidden list

File: torch/_inductor/ir.py [hunk 7]
```diff
@@ -3912,12 +4011,39 @@ def codegen_list_tuple_access(self, basename, indices):
...
+        for i, s in enumerate(self.get_size()):
+            if s in symbols_to_define:
+                wrapper.writeline(f"{s} = {self.get_name()}.size({i})")
+                symbols_to_define.remove(s)
+        for i, s in enumerate(self.get_stride()):
+            if s in symbols_to_define:
+                wrapper.writeline(f"{s} = {self.get_name()}.stride({i})")
+                symbols_to_define.remove(s)
+        if (s := self.get_offset()) in symbols_to_define:
+            wrapper.writeline(f"{s} = {self.get_name()}.storage_offset()")
```
Comment (aakhundov) [inline torch/_inductor/ir.py:4042]: Can we use `wrapper.declare` and `wrapper.ending` so that this also works in C++ wrapper codegen? Smth like this:

```
wrapper.writeline(f"{wrapper.declare}{s} = {self.get_name()}.size({i}){wrapper.ending}")
```

Comment (ezyang) [reply torch/_inductor/ir.py:4042]: i'm going to do this in a follow up, since there are some PRs backed up on this landing
